flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceSplitEnumerator.java

```diff
@@ -0,0 +1,406 @@
+        if (sourceEvent instanceof SourceReaderFinishedEvent) {
+            SourceReaderFinishedEvent srfe = (SourceReaderFinishedEvent) sourceEvent;
+            if (srfe.sourceIndex() != currentSourceIndex) {
+                if (srfe.sourceIndex() < currentSourceIndex) {
+                    // Assign pending splits if any
+                    TreeMap<Integer, List<HybridSourceSplit<SplitT>>> splitsBySource =
+                            pendingSplits.get(subtaskId);
+                    if (splitsBySource != null) {
+                        List<HybridSourceSplit<SplitT>> splits =
+                                splitsBySource.get(srfe.sourceIndex());
+                        if (splits != null && splits.isEmpty()) {
+                            // Splits have been processed by the reader
+                            splitsBySource.remove(srfe.sourceIndex());
+                        }
+                        if (splitsBySource.isEmpty()) {
+                            pendingSplits.remove(subtaskId);
+                        } else {
+                            Integer nextSubtaskSourceIndex = splitsBySource.firstKey();
+                            LOG.debug(
+                                    "Restore subtask={}, sourceIndex={}",
+                                    subtaskId,
+                                    nextSubtaskSourceIndex);
+                            context.sendEventToSourceReader(
+                                    subtaskId, new SwitchSourceEvent(nextSubtaskSourceIndex));
+                            assignPendingSplits(subtaskId);
+                        }
+                    }
+                    // Once all pending splits have been processed, add the readers to the current
+                    // enumerator, which may in turn trigger new split assignments
+                    if (!pendingReaders.isEmpty() && pendingSplits.isEmpty()) {
+                        // Advance pending readers to current enumerator
+                        LOG.debug(
+                                "Adding pending readers {} to enumerator currentSourceIndex={}",
+                                pendingReaders,
+                                currentSourceIndex);
+                        for (int pendingReaderSubtaskId : pendingReaders) {
+                            context.sendEventToSourceReader(
+                                    pendingReaderSubtaskId,
+                                    new SwitchSourceEvent(currentSourceIndex));
+                        }
+                        for (int pendingReaderSubtaskId : pendingReaders) {
+                            currentEnumerator.addReader(pendingReaderSubtaskId);
+                        }
+                        pendingReaders.clear();
+                    }
+                } else {
+                    // enumerator already switched
+                    LOG.debug("Ignoring out of order event {}", srfe);
+                }
+                return;
+            }
```

Comment:
this method is very long. maybe split it into smaller functions?

Also style is a little inconsistent. used `return` at the end of the `if` block. what about just traditional `if-else`? also it might make more sense if the if-else blocks are in separate methods

Reply:
This is now much simpler.